app/src/main/java/com/example/dashcarr/presentation/tabs/map/MapFragment.kt

```diff
@@ -271,215 +277,239 @@ class MapFragment : BaseFragment<FragmentMapBinding>(
-                magneticField,
-                SensorManager.SENSOR_DELAY_NORMAL,
-                SensorManager.SENSOR_DELAY_UI
-            )
+        viewLifecycleOwner.lifecycleScope.launch {
+            context?.let {
+                sensorRecordingViewModel.rpmLiveData.saveToFile(
+                    it,
+                    currentSensorStringObject,
+                    "filtered",
+                    "gyro",
+                    stopDateTime
+                )
+            }
         }
-    }
 
-    /**
-     * Stops listening for sensorChanges
-     */
-    override fun onPause() {
-        super.onPause()
-        sensorManager.unregisterListener(this)
-    }
+        viewLifecycleOwner.lifecycleScope.launch {
+            sensorRecordingViewModel.rpmLiveData.unfilteredGyroscopeList.collect {
+                currentSensorStringObject = sensorRecordingViewModel.rpmLiveData.buildSensorStringBuilder(it)
+            }
+        }
+        viewLifecycleOwner.lifecycleScope.launch {
+            context?.let {
+                sensorRecordingViewModel.rpmLiveData.saveToFile(
+                    it,
+                    currentSensorStringObject,
+                    "unfiltered",
+                    "gyro",
+                    stopDateTime
+                )
+            }
+        }
 
-    private fun startRecording() {
-        recordingViewModel.startRecording()
-        resetRecording()
-        isRecording = true
-        isTimerPaused = false
-        Toast.makeText(context, "Recording Started", Toast.LENGTH_SHORT).show()
-    }
+        viewLifecycleOwner.lifecycleScope.launch {
+            val existingJsonArray = context?.let { sensorRecordingViewModel.rpmLiveData.readJsonFromFile(it) }
+            context?.let {
+                if (existingJsonArray != null) {
+                    sensorRecordingViewModel.rpmLiveData.writeToJsonFile(it, existingJsonArray, allSensorJsonObject)
+                }
+            }
+        }
 
-    private fun stopRecording() {
-        recordingViewModel.stopRecording()
-        isRecording = false
-        saveToCSV()
+        Toast.makeText(context, "Recording Saved", Toast.LENGTH_SHORT).show()
     }
 
     private fun pauseRecording() {
         recordingViewModel.pauseRecording()
-        isRecording = false
-        isTimerPaused = true
+        viewLifecycleOwner.lifecycleScope.launch {
+            sensorRecordingViewModel.rpmLiveData.setIsRecording(false)
+
+        }
+        Toast.makeText(context, "Recording Paused", Toast.LENGTH_SHORT).show()
+
     }
 
     private fun resumeRecording() {
         recordingViewModel.resumeRecording()
-        isRecording = true
-        isTimerPaused = false
-        startTimeMillis = SystemClock.elapsedRealtime()
+        viewLifecycleOwner.lifecycleScope.launch {
+            sensorRecordingViewModel.rpmLiveData.setIsRecording(true)
+
+        }
+        Toast.makeText(context, "Recording Resumed", Toast.LENGTH_SHORT).show()
+
     }
```

Comment:
I am confused about the fact that you use two viewModels for the recording. For me, it seems like this is making stuff more complicated than it should be and it's not representing the domain because we only have one recording. [Take a short look into this.](https://en.wikipedia.org/wiki/Domain-driven_design) It might be a solution to make the sensorRecording viewModel into a normal class and use it in the recording view model.
This would also lead to many functions like this one to disappear.

In general let's talk about it. What were your thoughts? There might be a reason for it that I don't see. Feel free to call me today or tomorrow. You could also come over I noticed we are actually neighbors xD